Introduce VOMongoExecuteStrategy

tinchodias opened the pull request:
Introduce VOMongoExecuteStrategy, to help VOMongoRepositoryResolver in execution of persistence operations, which may need to retry depending on the error.

This new object centralizes error handling which is distributed in VOSessionPool, VOMongoSessionPool, and several methods in VOMongoRepositoryResolver. The idea to to make it more clear what happens with errors. 

Also, I added a couple of options in VOMongoRepository:
- executeRetryCount: the number of times to retry execution [it's now read-only option... we can add mutator very easily]
- debug: when true execution data is logged in transcript. Useful to understand what happens in some scenarios such as replication.

Notes:
- You may not like how VOMongoExecuteStrategy>>execute is written.
- I removed the infinite loop: now VOMongoSession does not handle the errors to reopen socket and retry.
- now VOSessionPool>>populate skips NetworkErrors. Please correct me if I'm wrong, but I think that's not necessary, and it's annoying to have the forked process opening many debuggers just after VOMongoRepository instance creation (when there is something wrong in the mongodb)

tinchodias commented:
Ping!

The goal of this change is to centralize the error handling; I needed to centralize them because when a replica-set is enabled, we need to pay attention to execution errors and act in consequence. For example, the error can tell you that the execution failed because the mongodb node was not primary. Then, our replica-set support acts in consequence by discovering who is the primary node, to retry the execution.

zecke commented on mc/Voyage-Mongo-Core.package/VOMongoExecuteStrategy.class/instance/execute.st:
```diff
@@ -0,0 +1,24 @@
+		do: [ :error | 
+			self logError: error.
+
+			(error isKindOf: MongoCommandError)
```
double dispatch?

tinchodias replied:
+1, good candidate

zecke commented on mc/Voyage-Mongo-Core.package/VOMongoExecuteStrategy.class/instance/retryOrResignal..st:
```diff
@@ -0,0 +1,10 @@
+retryOrResignal: error
+	"Execute again if there are still retries left. Else, just resignal the error."
+
+	^ retriesLeft > 0 
```
Nice to have the retries in a state/object now

tinchodias commented on mc/Voyage-Mongo-Core.package/VOMongoConnectionError.class/properties.json:
```diff
@@ -6,7 +6,7 @@
 	"commentStamp" : "<historical>",
 	"instvars" : [
-		 ],
+		"internalErrorsByTimestamp" ],
```
not sure if collecting internal errors is useful; it was intended for debugging reasons. It does not represent overhead when no errors, since it's lazily initialized, but it's not really needed. I can remove this feature in a new commit.

estebanlm commented:
@tinchodias you will implement the double-dispatch before I accept this PR?

tinchodias commented:
Yes

El 3/8/2016 7:40, "Esteban Lorenzano" <notifications@github.com> escribió:

> @tinchodias <https://github.com/tinchodias> you will implement the
> double-dispatch before I accept this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pharo-nosql/voyage/pull/47#issuecomment-237197626>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5zqfuyBlryZtvI5WUh7h0PAe0QfhFiks5qcGsvgaJpZM4JLDQm>
> .
>

tinchodias commented:
Buh... I still don't know how to use the filetree merge driver. In other cases I did workaround it manually: I filed out my changes and filed in in new head.

```
On branch ExecuteStrategy
Your branch is up-to-date with 'origin/ExecuteStrategy'.

You have unmerged paths.
  (fix conflicts and run "git commit")

Changes to be committed:

	modified:   mc/ConfigurationOfVoyageMongo.package/.filetree
	modified:   mc/ConfigurationOfVoyageMongo.package/ConfigurationOfVoyageMongo.class/instance/stable..st
	new file:   mc/ConfigurationOfVoyageMongo.package/ConfigurationOfVoyageMongo.class/instance/v1_3_7..st
	modified:   mc/ConfigurationOfVoyageMongo.package/ConfigurationOfVoyageMongo.class/methodProperties.json
	modified:   mc/ConfigurationOfVoyageMongo.package/ConfigurationOfVoyageMongo.class/properties.json
	modified:   mc/ConfigurationOfVoyageMongo.package/monticello.meta/version
	modified:   mc/ConfigurationOfVoyageMongo.package/properties.json
	new file:   mc/Voyage-Mongo-Core.package/VOLazyProxy.class/instance/becomeForwardKeepingHash..st
	modified:   mc/Voyage-Mongo-Core.package/VOLazyProxy.class/instance/becomeForwardWithSameShape..st
	modified:   mc/Voyage-Mongo-Core.package/VOLazyProxy.class/instance/doesNotUnderstand..st
	modified:   mc/Voyage-Mongo-Core.package/VOLazyProxy.class/methodProperties.json

Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both modified:      mc/Voyage-Mongo-Core.package/monticello.meta/version

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.gitattributes
```

and my .gitattributes contains:

```
*.package/monticello.meta/version       merge=mcVersion
*.package/*.class/methodProperties.json     merge=mcMethodProperties
*.package/*.class/properties.json       merge=mcProperties
*.package/*.extension/methodProperties.json     merge=mcMethodProperties
*.package/*.extension/properties.json       merge=mcProperties
```

estebanlm commented:
you have to add root path (from where repository starts):

    mc/*.package/monticello.meta/version       merge=mcVersion
    mc/*.package/*.class/methodProperties.json     merge=mcMethodProperties
    mc/*.package/*.class/properties.json       merge=mcProperties
    mc/*.package/*.extension/methodProperties.json     merge=mcMethodProperties
    mc/*.package/*.extension/properties.json       merge=mcProperties

tinchodias commented:
Now it works!!!! :-D thanks

On Mon, Aug 8, 2016 at 6:41 AM, Esteban Lorenzano <notifications@github.com>
wrote:

> you have to add root path (from where repository starts):
>
> mc/*.package/monticello.meta/version       merge=mcVersion
> mc/*.package/*.class/methodProperties.json     merge=mcMethodProperties
> mc/*.package/*.class/properties.json       merge=mcProperties
> mc/*.package/*.extension/methodProperties.json     merge=mcMethodProperties
> mc/*.package/*.extension/properties.json       merge=mcProperties
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pharo-nosql/voyage/pull/47#issuecomment-238188844>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5zqYiDSjb8C1akkMt8e9S9YK3-TGUfks5qdvm3gaJpZM4JLDQm>
> .
>

tinchodias commented on mc/Voyage-Mongo-Core.package/VOMongoExecuteStrategy.class/instance/execute.st:
```diff
@@ -0,0 +1,8 @@
+public
+execute
+
```
This is how it looks now, much better.

tinchodias commented:
I think now it can be merged

tinchodias commented:
Maybe I didn't say it but this PR is needed to adapt Voyage-Mongo to [new Mongo's error handling](https://github.com/pharo-nosql/mongotalk/pull/5) already merged in master.

estebanlm commented:
ok, ok... I will accept :)
